applications/HDF5Application/custom_io/hdf5_model_part_io.cpp

```diff
@@ -0,0 +1,265 @@
+#include "custom_io/hdf5_model_part_io.h"
+
+#include "includes/kratos_components.h"
+#include "custom_utilities/hdf5_points_data.h"
+#include "custom_utilities/hdf5_connectivities_data.h"
+#include "custom_utilities/hdf5_pointer_bins_utility.h"
+#include "custom_io/hdf5_properties_io.h"
+#include "custom_io/hdf5_nodal_solution_step_variables_io.h"
+#include "custom_io/hdf5_data_value_container_io.h"
+
+namespace Kratos
+{
+namespace HDF5
+{
+ModelPartIO::ModelPartIO(Parameters& rParams, File::Pointer pFile)
+: mpFile(pFile)
+{
+    KRATOS_TRY;
+
+    Parameters default_params(R"(
+        {
+            "prefix": "",
+            "list_of_elements": [],
+            "list_of_conditions": []
+        })");
```

Comment:
these `list` arguments are a bit unfortunate, since the mdpa should already have this information, just not in a convenient way (and I guess the hdf5 file input too).

What is the rationale for adding them? I would be in favor of parsing the input file once in advance to construct these lists if required.

Reply:
Initially the hdf5 input file doesn't exist. The problem I encountered was how to identify the element type in Kratos (these may not even be given in the mdpa). There is no function that returns the element's type (e.g., by string). If there were such a function, I could loop over all elements and conditions and identify the different types in the model part, but this would add computational overhead in exchange for user convenience. It would nevertheless simplify the function IsMatch() in the PointerBinsUtility. Currently, typeid by itself fails when the geometry is different but the element class is the same so I also have to check number of nodes and working space dimension.

Reply:
Ok, understood. You are right that the mdpa type is not necessarily the one you want. I would leave it as is for now, since there is no clean solution. However I think there is need in kratos of a functionality like you describe.

Thinking ahead, if I were to use the file you produce with hdf5 as input for a new simulation (say like a restart), could I avoid having to provide this string? I think having a way to do this will help us in the future.

Also, this is more of a wish, but one thing we want in the future is the ability to read pure geometries (not elements) and generate the elements depending on the solver. Unfortunaly we have no container for geometries on the model part, so it can't be done now. My question is: given the implementation of a container for lists of geometries, do you see your I/O as being able to work in this way?

Reply:
It is possible to automatically extract the element and condition types from the file directly for reading a model part. In python this can be done quite easily, in c++ we need to add the functionality to the file class. Currently, if you don't have the original json parameters you can also view the file hierarchy with hdfview and manually create the lists.

To your last question, I think it should be straightforward to extend. I think the bigger challenge is managing all the possible formats for io. Probably, it is easier to write a new class in this case. As it is now, there is not a lot of code in the model_part_io. Hopefully this makes it easier to write new types of io without having to worry so much about the low level stuff.